src/main/cpp/src/timezones.cu

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "timezones.hpp"
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/lists/list_device_view.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+
+using column = cudf::column;
+using column_device_view = cudf::column_device_view;
+using lists_column_device_view = cudf::detail::lists_column_device_view;
+using size_type = cudf::size_type;
+using struct_view = cudf::struct_view;
+
+namespace {
+
+/**
+ * @brief adjust timestamp value by offset
+ * 
+ * @tparam typestamp_type type of the input and output timestamp
+ * @param timestamp input timestamp
+ * @param transitions the transitions 
+ */
+template <typename timestamp_type>
+__device__ timestamp_type convert_timestamp_timezone(timestamp_type const& timestamp,
+    lists_column_device_view const& transitions,
```

Comment:
If you convert this to a struct/functor, you could eliminate the wrapping lambdas at the calls to thrust::transform and the counting iterator. eg:

```
template<typename timestamp_type>
struct convert_timestamp_timezone {
  list_column_device_view & fixed_transitions;
 size_type tz_index;
 

  __device__ timestamp_type operator()(timestamp_type t){
     ...
  }
};

thrust::transform(rmm::exec_policy(stream),
     input_data.begin<T>(),
     input_data.end<T>(),
     results->mutable_view().begin<cudf::timestamp_s>(),
     convert_timestamp_timezone{fixed_transitions, tz_index});
```

Reply:
I actually had this way before, but for some reason it was not the easiest code to debug. I couldn't step through the code or set breakpoints properly with (cuda-)gdb.

Reply:
Yeah, cuda-gdb falls apart pretty quickly.